### PR TITLE
Add command for downloading application package

### LINKED ISSUE #26487: New API to download application packages
From #26104

Build an API to get all files from an active application to redeply it incrementally

(I noticed vespa-deploy fetch, but could not find the same fetch in the APIs)

Use case:  an application package is already deployed with some schemas on a remote server or in the cloud. You want to add/remove/update a schema from it, without erasing other schemas. But you do not have access to the source of the last deployed files, nor to the server filesystem.

### PR DESCRIPTION
**Pull request by mpolden**
Fixes #26487.

@jonmv

### CODE REVIEW

File: client/go/internal/vespa/deploy.go [hunk 1]
```diff
@@ -119,6 +119,126 @@ func ZoneFromString(s string) (ZoneID, error) {
...
+	_, err = io.Copy(w, response.Body)
+	return err
+}
+
+func fetchFromController(deployment DeploymentOptions, path string) error {
+	var (
+		pkgURL *url.URL
+		err    error
+	)
+	if deployment.Target.Deployment().Zone.Environment == "dev" {
```
Comment (jonmv) [inline client/go/internal/vespa/deploy.go:154]: Or `perf`?

File: client/go/internal/vespa/deploy.go [hunk 2]
```diff
@@ -119,6 +119,126 @@ func ZoneFromString(s string) (ZoneID, error) {
...
+		pkgURL, err = deployment.url(fmt.Sprintf("/application/v4/tenant/%s/application/%s/instance/%s/job/%s/package",
+			deployment.Target.Deployment().Application.Tenant,
+			deployment.Target.Deployment().Application.Application,
+			deployment.Target.Deployment().Application.Instance,
+			deployment.Target.Deployment().Zone.Environment+"-"+deployment.Target.Deployment().Zone.Region,
+		))
+	} else {
+		pkgURL, err = deployment.url(fmt.Sprintf("/application/v4/tenant/%s/application/%s/package",
+			deployment.Target.Deployment().Application.Tenant,
+			deployment.Target.Deployment().Application.Application),
```
Comment (jonmv) [inline client/go/internal/vespa/deploy.go:164]: Not including the `build` option is fine with me here.

Comment (mpolden) [reply client/go/internal/vespa/deploy.go:164]: Agree, we can add if/when someone wants it.

File: client/go/internal/vespa/deploy.go [hunk 3]
```diff
@@ -119,6 +119,126 @@ func ZoneFromString(s string) (ZoneID, error) {
...
+			}
+		} else {
+			if err := os.MkdirAll(filepath.Dir(entryName), 0755); err != nil {
+				return err
+			}
+			f, err := os.Create(entryName)
+			if err != nil {
+				return err
+			}
+			if err := deployServiceGet(fu, deployment, f); err != nil {
```
Comment (jonmv) [inline client/go/internal/vespa/deploy.go:232]: Any reason not to `defer f.close()`?

**Review by jonmv — DISMISSED** [review]

File: client/go/internal/vespa/deploy.go (hunk 3 above)
Comment (mpolden) [reply client/go/internal/vespa/deploy.go:232]: This is called in a loop. If you defer closing, all files are kept open until the loop is done and the function returns. If the package has many files you could run into the open file limit.

Comment (jonmv) [reply client/go/internal/vespa/deploy.go:233]: Ah, thanks for educating me :)

**Review by jonmv — APPROVED** [review]

**Pull request merged**